js/utils/fees.js

```diff
@@ -9,41 +11,126 @@ export const feeLevels = [
+  }
+}
+
+/**
+ * Since a new transactions will potentially affect inputs and change the estimated fee,
+ * we'll clear the cache when any new transactions come.
+ */
+function watchTransactions() {
+  if (watchingTransactions) return;
+  watchingTransactions = true;
+
+  const socket = getSocket();
+  if (!socket) return;
+
+  socket.on('message', onSocket);
+
+  // in case we connect to a new server
+  serverConnectEvents.on('connected', e => {
+    socket.off(null, onSocket);
+    e.socket.on('message', onSocket);
+  });
+}
 
-export default function estimateFee(feeLevel) {
+/**
+ * Will estimate a fee based on the provided amount, fee level and the current state
+ * of the wallet.
+ *
+ * @param {string} feeLevel - The fee level
+ * @param {amount} number - The amount of the transaction in Bitcoin.
+ * @return {object} An jQuery defered which on success will resolve with the fee
+ *   in Bitcoin. If the call fails, the deferred will fail and pass on the args the
+ *   xhr fail handler receives.
+ */
+export function estimateFee(feeLevel, amount) {
   if (feeLevels.indexOf(feeLevel) === -1) {
     throw new Error(`feelevel must be one of ${feeLevels.join(', ')}`);
   }
 
+  if (typeof amount !== 'number') {
+    throw new Error('Please provide an amount as a number.');
+  }
+
+  watchTransactions();
+
   let deferred;
-  const cached = estimateFeeCache.get(feeLevel);
+  const cacheKey = `${feeLevel}-${amount}`;
+  const cached = estimateFeeCache.get(cacheKey);
 
   if (cached) {
     if (cached && Date.now() - cached.createdAt < cacheExpires) {
       deferred = cached.deferred;
     } else {
       // cache is expired
-      cached.deferred.reject();
+      estimateFeeCache.delete(cacheKey);
     }
   }
 
   if (!deferred) {
     deferred = $.Deferred();
 
-    estimateFeeCache.set(feeLevel, {
+    estimateFeeCache.set(cacheKey, {
       deferred,
       createdAt: Date.now(),
     });
+
+    const queryArgs = `feeLevel=${feeLevel}&amount=${decimalToInteger(amount, true)}`;
+    $.get(app.getServerUrl(`wallet/estimatefee/?${queryArgs}`))
+      .done((...args) => {
+        deferred.resolve(integerToDecimal(args[0].estimatedFee, true), ...args.slice(1));
+      }).fail((xhr, ...args) => {
+        deferred.reject(xhr, ...args);
+
+        const knownReasons = ['ERROR_INSUFFICIENT_FUNDS', 'ERROR_DUST_AMOUNT'];
+
+        // don't cache calls that failed with an unknown reason
+        if (xhr.responseJSON && knownReasons.indexOf(xhr.responseJSON.reason) === -1) {
+          estimateFeeCache.delete(cacheKey);
+        }
+      });
+  }
+
+  return deferred.promise();
+}
+
+let getFeesCache = {};
+
+/**
+ * Will call the fees api ('wallet/fees') on the server.
+ *
+ * @return {object} An jQuery defered which on success will resolve with the fees
+ *   per byte in Satoshi for each fee level. If the call fails, the deferred will
+ *   fail and pass on the args the xhr fail handler receives.
+ */
+export function getFees() {
```

Comment:
@jjeffryes Turns out I also needed to use `wallet/fees`, so I integrated it into the fees util module. It gives us caching which makes it easier to use it in certain use-cases. For example, let's say you have a list of transaction views and they all need to get the estimated fee. Instead of having to centralize it in a parent and pass it down, each child could just call the api knowing that the util module will batch them into one call.

It's probably not super urgent, but if you wanted to, feel free to plug this into the Purchase modal. Should be mostly as simple as replacing your `$.get()` call with a `getFees()` call.

Also, don't worry about aborting the cal. The returned promise doesn't expose an abort method. You wouldn't want to blindly abort, since another process may also be waiting on the call.